system/jlib/jmetrics.hpp

```diff
@@ -262,4 +273,42 @@ std::shared_ptr<CustomMetric<T>> createCustomMetricAndAddToReporter(const char *
     return pMetric;
 }
 
+
+class jlib_decl ScopedGaugeUpdater
+{
+public:
+    ScopedGaugeUpdater(const std::shared_ptr<GaugeMetric> &_pGauge, int64_t _amount=1):
+        pGauge{_pGauge},
+        amount{_amount}
+    {
+        pGauge->inc(amount);
+    }
+
+    ~ScopedGaugeUpdater()
+    {
+        pGauge->dec(amount);
+    }
+
+protected:
+    std::shared_ptr<GaugeMetric> pGauge;
```

Comment:
I'm concerned about the efficiency here.
If ScopedGaugeUpdater is used to track very rapid transactions, then the overhead of linking/using a shared_ptr is going to be more than the inc/dec itself.

Does it need to be a shared_ptr at this point? and/or can it just be a reference to one?

Reply:
@jakesmith 
I have been hesitant to store shared pointers as references since the reference does not increment the reference count. Should the owner of the scoped object decide to delete the object, the scoped object may fail when decrementing. I see some solutions:
1. use a reference and don't worry about it (i.e. trust that it won't be deleted)
2. Use 1, but add a null check prior to the decrement (if (pGauge) { do the decrement })
3. Live with the overhead (probably not acceptable)
4. Use a try-finally.

What do you think about (2)? It eliminates the copy and provides the safety in case the metric were deleted. I would also ensure the copy and assignment constructors are deleted for additional safety.

Reply:
The component created the metric and the component is using the guage updater to update it.
It needs a very low overhead way of doing so.
I think 1 is fine - the semantics are that during the live of the scoped object, we assert the metric is valid.
The componet/creator would have to invalidate it for that not to be so.

2. would still be additional wasted overhead, you'd need a weak pointer / lock to avoid potential race condition - but as per above - there shouldn't be one.

Reply:
I'll go with 1 and make the change. Since use of the class is the responsibility of the component, there is also no need to delete the copy and assignment constructors.